src/tools/wasm-ctor-eval.cpp

```diff
@@ -553,21 +564,222 @@ struct CtorEvalExternalInterface : EvallingModuleRunner::ExternalInterface {
+      // Add all other globals after them.
+      for (auto& global : oldGlobals) {
+        if (global) {
+          wasm->addGlobal(std::move(global));
+        }
+      }
+
+      // After sorting, perform the fixups that we need, that is, replace the
+      // relevant fields in cycles with a null and prepare a set in the start
+      // function.
+      //
+      // We'll track the set of readable globals as we go (which are the globals
+      // we've seen already, and fully finished processing).
+      std::unordered_set<Name> readableGlobals;
+
+      for (auto& global : wasm->globals) {
+        if (!global->init) {
+          continue;
+        }
+
+        struct InitFixer : PostWalker<InitFixer> {
+          CtorEvalExternalInterface& evaller;
+          std::unique_ptr<Global>& global;
+          std::unordered_set<Name>& readableGlobals;
+
+          InitFixer(CtorEvalExternalInterface& evaller,
+                    std::unique_ptr<Global>& global,
+                    std::unordered_set<Name>& readableGlobals)
+            : evaller(evaller), global(global),
+              readableGlobals(readableGlobals) {}
+
+          // Handles a child by fixing things up if needed. Returns true if we
+          // did in fact fix things up.
+          bool handleChild(Expression*& child,
+                           Expression* parent,
+                           Index fieldIndex = 0) {
+            if (!child) {
+              return false;
+            }
+
+            if (auto* get = child->dynCast<GlobalGet>()) {
+              if (!readableGlobals.count(get->name)) {
+                // This get cannot be read - it is a global that appears after
+                // us - and so we must fix it up, using the method mentioned
+                // before (setting it to null now, and later in the start
+                // function writing to it).
+                assert(isNullableAndMutable(parent, fieldIndex));
+                evaller.addStartSet(
+                  {global->name, global->type}, fieldIndex, get);
+                child =
+                  Builder(*getModule()).makeRefNull(get->type.getHeapType());
+                return true;
+              }
+            }
+
+            return false;
+          }
+
+          void visitStructNew(StructNew* curr) {
+            Index i = 0;
+            for (auto*& child : curr->operands) {
+              handleChild(child, curr, i++);
+            }
+          }
+          void visitArrayNew(ArrayNew* curr) {
+            if (handleChild(curr->init, curr)) {
+              // Handling array.new is tricky as the number of items may be
+              // unknown at compile time, so we'd need to loop at runtime. But,
+              // in practice we emit an array.new_fixed anyhow, so this should
+              // not be needed for now.
+              WASM_UNREACHABLE("TODO: ArrayNew in ctor-eval cycles");
+            }
+          }
+          void visitArrayNewFixed(ArrayNewFixed* curr) {
```

Comment:
Perhaps, but the abstraction we need here is "GC-creating instructions" which we don't have a utility for. That is, this isn't relevant for all constant instructions with a reference child, but only instructions that make GC data, because that is all we use for defining globals in the serialization. I'll add a comment.